### PR TITLE
Upper implementation and test

### PR DESCRIPTION
**Pull request by eric-feng-2011**
Upper implementation and test. This commit removes the docker readme, because git ...

### CODE REVIEW

File: docker/README.md [hunk 1]
```diff
@@ -1,20 +0,0 @@
-Docker configuration for testing and development of Opaque.
```
Comment (wzheng) [inline docker/README.md:1]: This diff shouldn't be here  since it's compared to the current master.

File: src/enclave/Enclave/ExpressionEvaluation.h [hunk 2]
```diff
@@ -961,6 +961,40 @@ class FlatbuffersExpressionEvaluator {
...
+
+      if (str->value_type() != tuix::FieldUnion_StringField) {
+        throw std::runtime_error(
+          std::string("tuix::Substring requires str String, not ")
+          + std::string("str ")
+          + std::string(tuix::EnumNameFieldUnion(str->value_type())));
+      }
+
+      // Obtain the input as a string
+      bool result_is_null = str->is_null();
```
Comment (wzheng) [inline src/enclave/Enclave/ExpressionEvaluation.h:977]: If the input field is NULL, then no further processing needs to be done and we can just return a NULL value.

File: src/main/scala/edu/berkeley/cs/rise/opaque/expressions/Upper.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+case class Upper(child: Expression)
+    extends UnaryExpression with String2StringExpression with NullIntolerant {
+
+  override def convert(v: UTF8String): UTF8String = v.toUpperCase
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    defineCodeGen(ctx, ev, c => s"($c).toUpperCase()")
+  }
+}
```
Comment (wzheng) [inline src/main/scala/edu/berkeley/cs/rise/opaque/expressions/Upper.scala:33]: This file doesn't need to be here since Upper is already defined in Catalyst.

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala [hunk 4]
```diff
@@ -582,6 +589,20 @@ trait OpaqueOperatorTests extends FunSuite with BeforeAndAfterAll { self =>
     df.select(dot($"v1", $"v2")).collect
   }
 
+  testAgainstSpark("upper") { securityLevel =>
```
Comment (wzheng) [inline src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala:592]: I'm curious how this test passes... could you paste the logical and physical plan queries here?

File: src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala [hunk 5]
```diff
@@ -106,6 +106,7 @@ import edu.berkeley.cs.rise.opaque.expressions.DotProduct
 import edu.berkeley.cs.rise.opaque.expressions.VectorAdd
 import edu.berkeley.cs.rise.opaque.expressions.VectorMultiply
 import edu.berkeley.cs.rise.opaque.expressions.VectorSum
+import edu.berkeley.cs.rise.opaque.expressions.Upper
```
Comment (wzheng) [inline src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala:109]: We should use Catalyst's Upper expression since it already exists. Upper shouldn't be a UDF.

**Review by wzheng — COMMENTED** [review]
Partial review.

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala (hunk 4 above)
Comment (wzheng) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala:592]: Also, the test should test for NULL values as well.

Comment (eric-feng-2011) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala:592]: scala> dfe.select(upper($"v1")).explain
== Physical Plan ==
EncryptedProject [upper(v1#4) AS upper(v1)#26]
+- EncryptedLocalTableScan [v1#4, v2#5], [[lower,upper]]

File: src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala (hunk 5 above)
Comment (eric-feng-2011) [reply src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala:109]: Changed to use catalyst upper

File: docker/README.md (hunk 1 above)
Comment (eric-feng-2011) [reply docker/README.md]: I think I got this resolved.

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala (hunk 4 above)
Comment (eric-feng-2011) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala:585]: Still working on adding the null test...

File: src/enclave/Enclave/ExpressionEvaluation.h (hunk 2 above)
Comment (eric-feng-2011) [reply src/enclave/Enclave/ExpressionEvaluation.h]: Added null parsing. Doesn't seem to be working with null test...

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala (hunk 4 above)
Comment (chester-leung) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala:585]: Maybe you can create two tests, one for nulls and one for non-nulls

File: src/enclave/Enclave/ExpressionEvaluation.h (hunk 2 above)
Comment (eric-feng-2011) [reply src/enclave/Enclave/ExpressionEvaluation.h]: Null test working. Null is null.asInstanceOf[String]

File: src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala (hunk 4 above)
Comment (eric-feng-2011) [reply src/test/scala/edu/berkeley/cs/rise/opaque/OpaqueOperatorTests.scala:591]: Added two tests - one including null and one without

**Review by wzheng — APPROVED** [review]
Looks good, thanks!
